packages/gatsby/src/schema/node-model.js

```diff
@@ -278,6 +483,254 @@ const toNodeTypeNames = (schema, gqlTypeName) => {
+        )
+      } else if (
+        isCompositeType(gqlFieldType) &&
+        _.isArray(innerValue) &&
+        gqlNonNullType instanceof GraphQLList
+      ) {
+        innerValue = await Promise.all(
+          innerValue.map(item =>
+            resolveRecursive(
+              nodeModel,
+              schemaComposer,
+              schema,
+              item,
+              gqlFieldType,
+              queryField,
+              _.isObject(fieldToResolve) ? fieldToResolve : queryField
+            )
+          )
+        )
+      }
+    }
+    if (innerValue != null) {
+      resolvedFields[fieldName] = innerValue
+    }
+  }
+
+  Object.keys(queryFields).forEach(key => {
+    if (!fieldsToResolve[key] && node[key]) {
+      resolvedFields[key] = node[key]
+    }
+  })
+
+  return _.pickBy(resolvedFields, (value, key) => queryFields[key])
+}
+
+function resolveField(
+  nodeModel,
+  schemaComposer,
+  schema,
+  node,
+  gqlField,
+  fieldName
+) {
+  const withResolverContext = require(`./context`)
+  return gqlField.resolve(
+    node,
+    gqlField.args.reduce((acc, arg) => {
+      acc[arg.name] = arg.defaultValue
+      return acc
+    }, {}),
+    withResolverContext({
+      schema,
+      schemaComposer,
+      nodeModel,
+    }),
+    {
+      fieldName,
+      schema,
+      returnType: gqlField.type,
+    }
+  )
+}
+
+const determineResolvableFields = (schemaComposer, schema, type, fields) => {
+  const fieldsToResolve = {}
+  const gqlFields = type.getFields()
+  Object.keys(fields).forEach(fieldName => {
+    const field = fields[fieldName]
+    const gqlField = gqlFields[fieldName]
+    const gqlFieldType = getNamedType(gqlField.type)
+    const typeComposer = schemaComposer.getAnyTC(type.name)
+    const needsResolve = typeComposer.getFieldExtension(
+      fieldName,
+      `needsResolve`
+    )
+    if (_.isObject(field) && gqlField) {
+      const innerResolved = determineResolvableFields(
+        schemaComposer,
+        schema,
+        gqlFieldType,
+        field
+      )
+      if (!_.isEmpty(innerResolved)) {
+        fieldsToResolve[fieldName] = innerResolved
+      } else if (_.isEmpty(innerResolved) && needsResolve) {
+        fieldsToResolve[fieldName] = true
+      }
+    } else if (needsResolve) {
+      fieldsToResolve[fieldName] = true
+    }
+  })
+  return fieldsToResolve
+}
+
+/**
+ * Add link between passed data and Node. This function shouldn't be used
+ * directly. Use higher level `trackInlineObjectsInRootNode`
+ * @see trackInlineObjectsInRootNode
+ * @param {(Object|Array)} data Inline object or array
+ * @param {string} nodeId Id of node that contains data passed in first parameter
+ */
+const addRootNodeToInlineObject = (
+  rootNodeMap,
+  data,
+  nodeId,
+  isNode = false
+) => {
+  const isPlainObject = _.isPlainObject(data)
+
+  if (isPlainObject || _.isArray(data)) {
+    _.each(data, (o, key) => {
+      if (isNode && key === `internal`) {
```

Comment:
we can drop the return 
```suggestion
      if (!isNode || key !== `internal`) {
        addRootNodeToInlineObject(rootNodeMap, o, nodeId)
      }
```